Tests/ApolloCodegenTests/CodeGenAST/IROperationBuilderTests.swift

```diff
@@ -0,0 +1,3031 @@
+  }
+
+  // MARK: - Children Computation
+
+  // MARK: Children - Fragment Type
+
+  /// Example:
+  /// query {
+  ///  allAnimals {
+  ///    ...AnimalDetails
+  ///  }
+  /// }
+  ///
+  /// fragment AnimalDetails on Animal {
+  ///   species
+  /// }
+  /// Expected:
+  /// Animal.children should not include a type case for asAnimal
+  func test__children__initWithNamedFragmentOnTheSameType_hasNoChildTypeCase() {
+    // given
+    let Interface_Animal = GraphQLInterfaceType.mock("Animal")
+
+    let animalDetails = CompilationResult.FragmentDefinition.mock("AnimalDetails", type: Interface_Animal)
+
+    operation = .mock(selections: [
+      .field(.mock(
+        "allAnimals",
+        selectionSet: .mock(
+          parentType: Interface_Animal,
+          selections: [
+            .fragmentSpread(animalDetails),
+          ]
+        )))])
+
+    // when
+    buildSubjectOperation()
+
+    let allAnimals = self.subject[field: "query"]?[field: "allAnimals"]?.selectionSet
+
+    // then
+    expect(allAnimals?.selections.typeCases).to(beEmpty())
+  }
+
+  /// Example:
+  /// query {
+  ///  allAnimals {
+  ///    ...BirdDetails
+  /// }
+  ///
+  /// fragment BirdDetails on Bird {
+  ///   species
+  /// }
+  /// Expected:
+  /// AllAnimals.children: [AsBird]
+  func test__children__initWithNamedFragmentOnMoreSpecificType_hasChildTypeCase() {
+    // given
+    let Interface_Animal = GraphQLInterfaceType.mock("Animal")
+    let Object_Bird = GraphQLObjectType.mock("Bird")
+    
+    let birdDetails = CompilationResult.FragmentDefinition.mock("BirdDetails", type: Object_Bird)
+
+    operation = .mock(selections: [
+      .field(.mock(
+        "allAnimals",
+        selectionSet: .mock(
+          parentType: Interface_Animal,
+          selections: [
+            .fragmentSpread(birdDetails),
+          ]
+        )))])
+
+    // when
+    buildSubjectOperation()
+
+    let allAnimals = self.subject[field: "query"]?[field: "allAnimals"]?.selectionSet
+
+    // then
+    expect(allAnimals?.selections.typeCases.count).to(equal(1))
+
+    let child = allAnimals?[as: "Bird"]
+    expect(child?.parentType).to(equal(Object_Bird))
+    expect(child?.selections.fragments).to(shallowlyMatch([birdDetails]))
+  }
+
+  /// Example:
+  /// query {
+  ///  birds {
+  ///    ...AnimalDetails
+  /// }
+  ///
+  /// fragment AnimalDetails on Animal {
+  ///   species
+  /// }
+  /// Expected:
+  /// Children should not include a type case for asAnimal
+  func test__children__isObjectType_initWithNamedFragmentOnLessSpecificMatchingType_hasNoChildTypeCase() {
+    // given
+    let Interface_Animal = GraphQLInterfaceType.mock("Animal")
+    let Object_Bird = GraphQLObjectType.mock("Bird", interfaces: [Interface_Animal])
+    let animalDetails = CompilationResult.FragmentDefinition.mock("AnimalDetails", type: Interface_Animal)
+
+    operation = .mock(selections: [
+      .field(.mock(
+        "allAnimals",
```

Comment:
Should this be `"birds"`? `[field: "allAnimals"]` would need to change down below too.